File: python/mirage/core/github/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import logging

from mirage.accessor.github import GitHubAccessor
from mirage.cache.index import (NULL_INDEX, IndexCacheStore, IndexEntry,
                                LookupStatus)
from mirage.cache.index.lock import index_lock
from mirage.core.github.repo import ensure_ref
from mirage.core.github.tree import (ensure_live_index, fetch_dir_tree,
                                     refill_index)
from mirage.core.github.tree_entry import TreeEntry
from mirage.types import PathSpec
from mirage.utils.errors import enoent
from mirage.utils.key_prefix import mount_prefix_of

log = logging.getLogger(__name__)


async def readdir(
    accessor: GitHubAccessor,
    path_spec: PathSpec,
    index: IndexCacheStore = NULL_INDEX,
) -> list[str]:
    prefix = mount_prefix_of(path_spec.virtual, path_spec.vfs_path)
    async with index_lock(index, prefix.rstrip("/") or "/"):
        return await _readdir(accessor, path_spec, index)


async def _readdir(
    accessor: GitHubAccessor,
    path_spec: PathSpec,
    index: IndexCacheStore = NULL_INDEX,
) -> list[str]:
    """Read while the caller holds the mount's index lock through lookup."""
    virtual = path_spec.virtual
    prefix = mount_prefix_of(path_spec.virtual, path_spec.vfs_path)
    path = (path_spec.dir if path_spec.pattern else path_spec).mount_path
    key = path.strip("/")
    virtual_key = prefix + "/" + key if key else prefix or "/"
    await ensure_live_index(accessor, index, prefix)
    listing = await index.list_dir(virtual_key)
    # The index is the whole listing here, not a cache in front of one, so
    # an *expired* answer means the tree aged out, not that the path is
    # gone. Refetch once and ask again. A NOT_FOUND against a live index
    # is a real absence and must not cost a tree fetch.
    if listing.status == LookupStatus.EXPIRED and not accessor.truncated:
        if await refill_index(accessor, index, prefix):
            listing = await index.list_dir(virtual_key)
    if listing.entries is not None:
        return listing.entries
    if accessor.truncated and listing.status in (LookupStatus.NOT_FOUND,
                                                 LookupStatus.EXPIRED):
        return await _fallback_readdir(accessor, virtual_key, index, virtual,
                                       prefix)
    if listing.status == LookupStatus.NOT_FOUND:
        raise enoent(virtual)
    return []


async def _fallback_readdir(
    accessor: GitHubAccessor,
    virtual_key: str,
    index: IndexCacheStore,
    virtual: str,
    prefix: str,
) -> list[str]:
    """Per-directory tree fetch when recursive tree was truncated.

    Args:
        accessor (GitHubAccessor): backend handle.
        virtual_key (str): Mount-absolute directory key.
        index (IndexCacheStore): the mount's index.
        virtual (str): Virtual path, for the error message.
        prefix (str): the mount prefix, which the descent has to strip to
            walk the repository's own path segments.
    """
    parent_sha = await _resolve_dir_sha(accessor, virtual_key, index, prefix)
    if parent_sha is None:
        raise enoent(virtual)
    entries = await fetch_dir_tree(accessor.config, accessor.owner,
                                   accessor.repo, parent_sha, accessor.pool)
    return await _cache_dir(index, virtual_key, entries)


async def _cache_dir(index: IndexCacheStore, virtual_key: str,
                     entries: list[TreeEntry]) -> list[str]:
    """Cache one complete tree listing, including each traversed parent."""
    norm = virtual_key.rstrip("/") or "/"
    child_keys: list[str] = []
    dir_entries: list[tuple[str, IndexEntry]] = []
    for entry in entries:
        child_path = norm.rstrip("/") + "/" + entry.path
        resource_type = "folder" if entry.type == "tree" else "file"
        idx_entry = IndexEntry(
            id=entry.sha,
            name=entry.path,
            resource_type=resource_type,
            size=entry.size,
        )
        dir_entries.append((entry.path, idx_entry))
        child_keys.append(child_path)
    await index.set_dir(norm, dir_entries)
    log.debug("fallback readdir populated %d entries for %s", len(entries),
              virtual_key)
    return sorted(child_keys)


async def _resolve_dir_sha(
    accessor: GitHubAccessor,
    virtual_key: str,
    index: IndexCacheStore,
    prefix: str = '',
) -> str | None:
    """Get the tree SHA for a directory path.

    Walks from the current ref, fetching per-directory trees. A cached
    entry may name an old tree even when its parent's listing is fresh.

    Args:
        accessor (GitHubAccessor): backend handle.
        virtual_key (str): Mount-absolute directory key.
        index (IndexCacheStore): the mount's index.
        prefix (str): the mount prefix the keys are built against.
    """
    norm = virtual_key.rstrip("/") or "/"
    stem = prefix.rstrip("/")
    rest = norm[len(stem):] if stem and norm.startswith(stem) else norm
    parts = [p for p in rest.strip("/").split("/") if p]
    current_sha = await ensure_ref(accessor)
    current_path = stem or "/"
    for part in parts:
        entries = await fetch_dir_tree(accessor.config, accessor.owner,
                                       accessor.repo, current_sha,
                                       accessor.pool)
        child_path = current_path.rstrip("/") + "/" + part
        found = next((entry for entry in entries if entry.path == part), None)
        if found is None or found.type != "tree":
            # Remove the former directory before caching a replacement blob.
            await index.invalidate_prefix(child_path)
        await _cache_dir(index, current_path, entries)
        if found is None or found.type != "tree":
            return None
        current_sha = found.sha
        current_path = child_path
    return current_sha

File: typescript/packages/core/src/core/github/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { mountPrefixOf } from '../../utils/key_prefix.ts'
import type { GitHubAccessor } from '../../accessor/github.ts'
import { LookupStatus } from '../../cache/index/config.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { fetchDirTree, type GitHubTreeItem } from './client.ts'
import { ensureLiveIndex, refillIndex } from './tree.ts'
import { withIndexLock } from '../../cache/index/lock.ts'
import { IndexEntry } from '../../cache/index/config.ts'
import { rstripSlash, stripSlash } from '../../utils/slash.ts'
import { enoent } from '../../utils/errors.ts'
import { compareCodePoints } from '../../utils/sort.ts'

function stripPrefix(path: PathSpec): string {
  const prefix = mountPrefixOf(path.virtual, path.vfsPath)
  let p = path.pattern !== null ? path.directory : path.virtual
  if (prefix !== '' && p.startsWith(prefix)) {
    p = p.slice(prefix.length) || '/'
  }
  return p
}

export async function readdir(
  accessor: GitHubAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  if (index === undefined) return readdirUnlocked(accessor, path, index)
  const prefix = mountPrefixOf(path.virtual, path.vfsPath)
  return withIndexLock(index, rstripSlash(prefix) || '/', () =>
    readdirUnlocked(accessor, path, index),
  )
}

/** Caller holds the mount's index lock through its final lookup. */
export async function readdirUnlocked(
  accessor: GitHubAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  if (index === undefined) {
    throw enoent(path.virtual)
  }
  const prefix = mountPrefixOf(path.virtual, path.vfsPath)
  const rel = stripSlash(stripPrefix(path))
  const key =
    rel === '' ? (prefix === '' ? '/' : rstripSlash(prefix)) : `${rstripSlash(prefix)}/${rel}`

  await ensureLiveIndex(accessor, index, prefix)
  let listing = await index.listDir(key)
  // The index is the whole listing here, not a cache in front of one, so an
  // *expired* answer means the tree aged out, not that the path is gone.
  // Refetch once and ask again. A NOT_FOUND against a live index is a real
  // absence and must not cost a tree fetch: refilling on any miss spends a
  // recursive-tree call on every ENOENT.
  if (listing.status === LookupStatus.EXPIRED && !accessor.truncated) {
    if (await refillIndex(accessor, index, prefix)) listing = await index.listDir(key)
  }
  if (listing.entries !== undefined && listing.entries !== null) {
    return listing.entries
  }
  if (
    accessor.truncated &&
    (listing.status === LookupStatus.NOT_FOUND || listing.status === LookupStatus.EXPIRED)
  ) {
    return fallbackReaddir(accessor, key, index, prefix)
  }
  if (listing.status === LookupStatus.NOT_FOUND) throw enoent(path)
  return []
}

async function fallbackReaddir(
  accessor: GitHubAccessor,
  key: string,
  index: IndexCacheStore,
  prefix: string,
): Promise<string[]> {
  const parentSha = await resolveDirSha(accessor, key, index, prefix)
  if (parentSha === null) throw enoent(key)
  const entries = await fetchDirTree(accessor.transport, accessor.owner, accessor.repo, parentSha)
  return cacheDir(index, key, entries)
}

// Cache one complete tree listing, including each traversed parent.
async function cacheDir(
  index: IndexCacheStore,
  key: string,
  entries: GitHubTreeItem[],
): Promise<string[]> {
  const childKeys: string[] = []
  const childEntries: [string, IndexEntry][] = []
  for (const e of entries) {
    const childKey = `${key === '/' ? '' : key}/${e.path}`
    childKeys.push(childKey)
    // setDir composes the child key as parent + '/' + name, so the tuple
    // carries the bare entry name, not the full path (mirrors python).
    childEntries.push([
      e.path,
      new IndexEntry({
        id: e.sha,
        name: e.path,
        vfsName: e.path,
        resourceType: e.type === 'tree' ? 'folder' : 'file',
        size: e.size ?? null,
      }),
    ])
  }
  childKeys.sort(compareCodePoints)
  await index.setDir(key, childEntries)
  return childKeys
}

async function resolveDirSha(
  accessor: GitHubAccessor,
  key: string,
  index: IndexCacheStore,
  prefix: string,
): Promise<string | null> {
  // Cached directory SHAs may belong to an older branch head, even when
  // their parent listing is still fresh. Resolve the path from the ref.
  const stem = rstripSlash(prefix)
  const rest = stem !== '' && key.startsWith(stem) ? key.slice(stem.length) : key
  const parts = stripSlash(rest)
    .split('/')
    .filter((p) => p !== '')
  let currentSha = accessor.ref
  let currentPath = stem || '/'
  for (const part of parts) {
    const entries = await fetchDirTree(
      accessor.transport,
      accessor.owner,
      accessor.repo,
      currentSha,
    )
    const childPath = `${currentPath === '/' ? '' : currentPath}/${part}`
    const found = entries.find((e) => e.path === part)
    if (found?.type !== 'tree') {
      // Remove the former directory before caching a replacement blob.
      await index.invalidatePrefix(childPath)
    }
    await cacheDir(index, currentPath, entries)
    if (found?.type !== 'tree') return null
    currentSha = found.sha
    currentPath = childPath
  }
  return currentSha
}
